compyle/profile.py

```diff
@@ -0,0 +1,100 @@
+""" Utils for profiling kernels
+"""
+
+import time
+from contextlib import contextmanager
+from collections import defaultdict
+from .config import get_config
+
+
+_profile_info = defaultdict(lambda: {'calls': 0, 'time': 0})
+
+
+def record_profile(name, time):
+    global _profile_info
+    _profile_info[name]['time'] += time
+    _profile_info[name]['calls'] += 1
+
+
+@contextmanager
+def profile(name):
+    start = time.time()
+    yield start
+    end = time.time()
+    record_profile(name, end - start)
+
+
+def profile_method(method):
+    def wrapper(*args, **kwargs):
+        self = args[0] if len(args) else None
+        with profile(getattr(self, "name", method.__name__)):
+            return method(*args, **kwargs)
+    return wrapper
+
+
+def print_profile():
+    global _profile_info
+    profile_data = sorted(_profile_info.items(), key=lambda x: x[1]['time'],
+                          reverse=True)
+    if len(_profile_info) == 0:
+        print("No profiling information available")
+        return
+    print("Profiling info:")
+    print("{:<40} {:<10} {:<10}".format('Function', 'N calls', 'Time'))
+    tot_time = 0
+    for kernel, data in profile_data:
+        print("{:<40} {:<10} {:<10}".format(
+            kernel,
+            data['calls'],
+            data['time']))
+        tot_time += data['time']
+    print("Total profiled time: %g secs" % tot_time)
+
+
+def profile_kernel(kernel, name, backend=None):
+    from compyle.array import get_backend
+    backend = get_backend(backend)
+
+    def _profile_knl(*args, **kwargs):
+        if backend == 'opencl':
+            event = kernel(*args, **kwargs)
+            event.wait()
+            exec_time = (event.profile.end - event.profile.start) * 1e-9
+            record_profile(name, exec_time)
+            return event
+        elif backend == 'cuda':
+            exec_time = kernel(*args, **kwargs, time_kernel=True)
+            record_profile(name, exec_time)
+            return exec_time
+        else:
+            start = time.time()
+            kernel(*args, **kwargs)
+            end = time.time()
+            record_profile(name, end - start)
+
+    if get_config().profile:
+        wgi = getattr(kernel, 'get_work_group_info', None)
+        if wgi is not None:
+            _profile_knl.get_work_group_info = wgi
+        return _profile_knl
+    else:
+        return kernel
+
+
+def named_profile(name, backend=None):
```

Comment:
This doesn't seem to be used anywhere?  Or is it only used in pysph?  Perhaps some tests are in order in that case as there seem to be no tests for any of this.  Sorry I am hassling you.

Reply:
This is only used in pysph. I have added some tests for these

Reply:
Great, thanks.